File: kernel/internal/pipeline/pipeline.go
package pipeline

import (
	"context"

	"agentos/kernel/internal/eventbus"
	"agentos/kernel/internal/session"
	"agentos/kernel/internal/tools"
)

// Response 是 Pipeline 一次调用的结果。
type Response struct {
	Allowed bool
	Errored bool
	Message string
	Result  map[string]any
}

// Pipeline 是 6 步统一管道。Kernel 核心只调它，不认识任何具体工具。
type Pipeline struct {
	registry *tools.Registry
	bus      eventbus.Bus
}

func New(registry *tools.Registry, bus eventbus.Bus) *Pipeline {
	return &Pipeline{registry: registry, bus: bus}
}

// Call 执行一次工具调用，经过：查找 → 提取权限 → 权限检查 → 资源扣减 → 执行 → 脱敏 → 审计（事件）。
func (p *Pipeline) Call(sess *session.Session, toolName string, params map[string]any) Response {
	ctx := context.Background()

	// 步骤 1：查找工具。未知工具 = 拒绝（不泄露工具清单细节）。
	tool, ok := p.registry.Get(toolName)
	if !ok {
		p.bus.Publish(eventbus.Event{
			Type: "tool.denied", SessionID: sess.ID, Tool: toolName, Params: params,
		})
		return Response{Allowed: false, Message: "permission denied"}
	}

	// 步骤 2：提取权限资源（工具自描述）。
	res := tool.PermissionKey(params)

	// 步骤 3：权限检查。不通过 → 拒绝 + 审计（拒绝原因不回传给调用方，防泄露）。
	if !sess.Gate.Allowed(toolName, res) {
		p.bus.Publish(eventbus.Event{
			Type: "tool.denied", SessionID: sess.ID, Tool: toolName, Params: params,
		})
		return Response{Allowed: false, Message: "permission denied"}
	}

	// 步骤 3.5：资源扣减。超限 → 硬终止 + quota.exceeded 事件。
	if err := sess.Account.Charge(session.Usage{Steps: 1}); err != nil {
		p.bus.Publish(eventbus.Event{
			Type: "quota.exceeded", SessionID: sess.ID, Tool: toolName, Params: params,
		})
		return Response{Errored: true, Message: "quota exceeded"}
	}

	// 步骤 4：执行工具。
	result, err := tool.Execute(ctx, params)
	if err != nil {
		p.bus.Publish(eventbus.Event{
			Type: "tool.errored", SessionID: sess.ID, Tool: toolName, Params: params,
		})
		return Response{Errored: true, Message: "tool error"}
	}

	// 步骤 5：脱敏（第一道安全防线，独立于工具）。保留摘要进事件。
	var sanSummary []eventbus.FieldSanitization
	if sess.Sanitizer != nil {
		sr := sess.Sanitizer.Sanitize(result.Data)
		result.Data = sr.Data
		// 摘要类型转换：sanitize.FieldSanitization → eventbus.FieldSanitization（解耦）。
		for _, f := range sr.Summary {
			sanSummary = append(sanSummary, eventbus.FieldSanitization{Field: f.Field, Strategy: f.Strategy})
		}
	}

	// 步骤 6：审计（通过事件触发 AuditSubscriber，不散落）。含脱敏摘要。
	p.bus.Publish(eventbus.Event{
		Type:     "tool.called", SessionID: sess.ID, Tool: toolName,
		Params: params, Result: result.Data, Sanitize: sanSummary,
	})

	return Response{Allowed: true, Result: result.Data}
}

File: cp/pipeline/pipeline.py
from dataclasses import dataclass, field
from typing import Any, Dict

from cp.eventbus.bus import Bus, Event, FieldSanitization
from cp.session.account import QuotaExceeded, Usage
from cp.session.session import Session
from cp.tools.tool import Registry


@dataclass
class PipelineResponse:
    allowed: bool = False
    errored: bool = False
    message: str = ""
    result: Dict[str, Any] = field(default_factory=dict)


class Pipeline:
    """6 步统一管道(async)。控制面核心只调它,不认识任何具体工具。

    步骤:查找 → 提取权限 → 权限检查 → 资源扣减 → 执行(经 SandboxPort)→ 脱敏 → 审计(事件)。
    编排器不直接 execute(ch15 约束7);执行委托给 SandboxPort。
    """

    def __init__(self, registry: Registry, bus: Bus, sandbox):
        self.registry = registry
        self.bus = bus
        self.sandbox = sandbox

    async def call(self, sess: Session, sandbox_id: str, tool_name: str, params: Dict[str, Any]) -> PipelineResponse:
        # 步骤 1:查找工具。未知工具 = 拒绝。
        tool, ok = self.registry.get(tool_name)
        if not ok:
            await self.bus.publish(Event(type="tool.denied", session_id=sess.id, tool=tool_name, params=params))
            return PipelineResponse(allowed=False, message="permission denied")

        # 步骤 2:提取权限资源(纯计算,不需 sandbox)。
        res = tool.permission_key(params)

        # 步骤 3:权限检查。
        if not sess.gate.allowed(tool_name, res):
            await self.bus.publish(Event(type="tool.denied", session_id=sess.id, tool=tool_name, params=params))
            return PipelineResponse(allowed=False, message="permission denied")

        # 步骤 3.5:资源扣减。
        try:
            await sess.account.charge(Usage(steps=1))
        except QuotaExceeded:
            await self.bus.publish(Event(type="quota.exceeded", session_id=sess.id, tool=tool_name, params=params))
            return PipelineResponse(errored=True, message="quota exceeded")

        # 步骤 4:执行——经 SandboxPort(不直接 tool.execute!)约束7
        action = {"tool": tool_name, "params": params}
        exec_result = await self.sandbox.exec_action(sandbox_id, action)
        if "error" in exec_result:
            await self.bus.publish(Event(type="tool.errored", session_id=sess.id, tool=tool_name, params=params))
            return PipelineResponse(errored=True, message="tool error")
        result_data = exec_result["data"]

        # 步骤 5:脱敏。
        san_summary = []
        if sess.sanitizer is not None:
            sr = sess.sanitizer.sanitize(result_data)
            san_summary = [FieldSanitization(field=f.field, strategy=f.strategy) for f in sr.summary]
            result_data = sr.data

        # 步骤 6:审计事件(含脱敏摘要)。
        await self.bus.publish(Event(
            type="tool.called", session_id=sess.id, tool=tool_name,
            params=params, result=result_data, sanitize=san_summary,
        ))

        return PipelineResponse(allowed=True, result=result_data)
